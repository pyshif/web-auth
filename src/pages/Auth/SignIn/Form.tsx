import styled from 'styled-components';
import { useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import Sep from 'components/Sep';
import { Form as F, Checkbox, Button, Input, message } from 'antd';
import { LockOutlined, UserOutlined } from '@ant-design/icons';
import routes from 'utils/routes';
import Link from 'components/Link';
import { useAppSelector, useAppDispatch } from 'store';
import { apiSignIn, DataSignIn } from 'store/features/authSlice';

const SignInForm = styled(F)`
    width: 100%;
    max-width: 360px;
    border: 1px solid rgba(240, 240, 240);
    border-radius: 0.5rem;
    padding: 2.875rem 2.5rem;
    margin: auto;
`;

const SubmitButton = styled(Button).attrs({
    htmlType: 'submit',
})`
    width: 100%;
`;

const GoogleButton = styled(Button)`
    width: 100%;
`;

function Form() {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();
    const { status, token, user } = useAppSelector((state) => state.auth);

    async function onFinish(values: any) {
        // console.log('values :>> ', values);
        const { email, password, remember } = values;

        // payload
        const data: DataSignIn = {
            email,
            password,
        };

        // request
        const hide = message.loading('Sign-in in progress...', 0);
        dispatch(apiSignIn(data)).then((action) => {
            const { error } = action as unknown as any;
            if (error) {
                message.error('Sign-in failed!', 3);
                return hide();
            }
            message.success('Sign-in success!', 3);
            navigate(routes.user);
            // remember feature
            // FIXME: protect user data in localstorage
            // cookies -> aes -> user email, password
            if (remember) {
                localStorage.setItem(
                    'C4UR',
                    JSON.stringify({ email, password })
                );
            } else {
                localStorage.removeItem('C4UR');
            }
            return hide();
        });
    }

    function onGoogleButton(e: any) {
        console.log('Received values of form :>> ', e);
        // TODO: complete google login in

        // load google gis library -> define loading handler
    }

    let localUser = {
        email: '',
        password: '',
    };
    if (localStorage.getItem('C4UR')) {
        localUser = JSON.parse(localStorage.getItem('C4UR') as string);
    }

    return (
        <SignInForm
            name="auth-signin"
            initialValues={{
                email: localUser.email,
                password: localUser.password,
                remember: false,
            }}
            onFinish={onFinish}
        >
            <F.Item
                name="email"
                rules={[
                    {
                        required: true,
                        message: 'Please input your email!',
                    },
                    {
                        type: 'email',
                        message: 'Please input correct email fromat!',
                    },
                ]}
            >
                <Input
                    prefix={<UserOutlined className="site-form-item.icon" />}
                    placeholder="Email"
                />
            </F.Item>
            <F.Item
                name="password"
                rules={[
                    {
                        required: true,
                        message: 'Please input your password!',
                    },
                ]}
            >
                <Input
                    prefix={<LockOutlined className="site-form-item-icon" />}
                    type="password"
                    placeholder="Password"
                />
            </F.Item>
            <F.Item>
                <F.Item
                    name="remember"
                    valuePropName="checked"
                    wrapperCol={{ span: 12 }}
                    noStyle
                >
                    <Checkbox>Remember me</Checkbox>
                </F.Item>
                <Link
                    to={routes.auth.forgot}
                    className="float-right"
                    scrollToTop
                >
                    Forgot password ?
                </Link>
            </F.Item>

            <F.Item>
                <SubmitButton>Log in</SubmitButton>
            </F.Item>
            <Sep />
            <F.Item>
                <GoogleButton onClick={onGoogleButton}>Googole</GoogleButton>
            </F.Item>
            <F.Item className="float-right">
                Or{' '}
                <Link
                    to={routes.auth.signup}
                    className="after:content-['_ðŸ‘‰']"
                    scrollToTop
                >
                    {' '}
                    Sign up by here
                </Link>
            </F.Item>
        </SignInForm>
    );
}

export default Form;
