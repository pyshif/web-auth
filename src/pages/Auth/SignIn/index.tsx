import styled from 'styled-components';
import { Card, Image } from 'antd';
import routes from 'utils/routes';
import logo from 'images/company-brand.png';
import Link from 'components/Link';
import Form from './Form';
import { font } from 'utils/font';

// Card
// Login Panel

// TODO: 程式設計思考：
// 1. 演算法：輸入 -> .... -> 輸出
// 2. 角度一：生產者 -> .... -> 消費者
// 3. 角度二：底層 -> .... -> 上層
// 4. 角度三：材料 -> .... -> 商品
// 5. 角度 N： ...

// 換句話說，當我們在編寫時，要思考自己在整個程式的哪個架構；
// 意思就是，我們有什麼材料、資料可以拿，我們想要產出什麼？
// 如果當我們，材料、資料取得後，到要產出的結果，中間處理非常繁雜時，就要將問題拆分
// 意思就是，你的材料、資料，太粗糙，需要先進行初步加工，在後面的製程才能更順利
// 這個時候就要思考，每一次的加工，要達到什麼樣的效果？能夠讓下一階段更好的使用？
// 通常這個過程就稱之為『優化』，然而『大量抽象性的思考通常不是很容易，也容易離實際應用場景背離』
// 所以『優化』這個過程是反覆的、重構的

// styled component 的材料是 網頁上各種 HTML Element 服務的是 React 基礎的元件
// 那麼當你需要組合出頁面時，中間可能還會需要經過自己產出的元件（加工過程），
// 這時就要思考這一階段要產出什麼，達到什麼效果？

function SignIn({ className }: { className: string }) {
    return (
        <>
            <LogoCard />
            <Form />
        </>
        // <div>
        //     <div className="w-16 mx-auto mb-6">
        //         <img src={Logo} alt="" />
        //     </div>
        //     <h1 className="title font-medium text-3xl text-center mb-3">
        //         Sign in to your account
        //     </h1>
        //     <h2 className="sub-titile text-md text-center mb-6">
        //         Or{' '}
        //         <span className="text-slate-600">
        //             you can{' '}
        //             <a href={routes.auth.signup} className="underline">
        //                 sign up by here
        //             </a>
        //         </span>
        //     </h2>
        // </div>
    );
}

function OldSignIn() {
    return (
        <div className="w-full max-w-sm mx-auto">
            {/* HEADER */}
            <div>
                <div className="w-16 mx-auto mb-6">
                    <img src={Logo} alt="" />
                </div>
                <h1 className="title font-medium text-3xl text-center mb-3">
                    Sign in to your account
                </h1>
                <h2 className="sub-titile text-md text-center mb-6">
                    Or{' '}
                    <span className="text-slate-600">
                        you can{' '}
                        <a href={routes.auth.signup} className="underline">
                            sign up by here
                        </a>
                    </span>
                </h2>
            </div>
            {/* FORM */}
            <div className="shadow-xl bg-white rounded-xl p-8">
                <form>
                    <div className="mb-3">
                        <label
                            htmlFor="email-address"
                            className="block indent-1 mb-1"
                        >
                            Email address
                        </label>
                        <input
                            type="email"
                            name="emailAddress"
                            id="email-address"
                            className="w-full h-8 px-1 border-[1px] rounded-md"
                        />
                        <div id="email-error-message"></div>
                    </div>
                    <div className="mb-3">
                        <label
                            htmlFor="password"
                            className="block indent-1 mb-1"
                        >
                            Password
                        </label>
                        <input
                            type="password"
                            name="password"
                            id="password"
                            className="w-full h-8 px-1 border-[1px] rounded-md"
                        />
                        <div id="password-error-message"></div>
                    </div>
                    {/* REMEMBER ME */}
                    {/* FORGOT PASSOWRD */}
                    <div className="mb-3 flow-root">
                        <a
                            href={routes.auth.forgot}
                            className="text-sm text-slate-800 float-right"
                        >
                            Forgot your password ?
                        </a>
                    </div>
                    {/* SIGN IN */}
                    <div className="mb-3 text-center">
                        <button
                            className="text-xl w-full h-9 bg-gray-700 text-white rounded-md"
                            onClick={(e) => {
                                e.preventDefault();
                                console.log(123);
                            }}
                        >
                            Sign in
                        </button>
                    </div>
                </form>
                <div className="relative flex mb-3">
                    <div className="grow translate-y-1/2 border-t-[1px] border-slate-400">
                        {' '}
                    </div>
                    <div className="whitespace-normal"> Or continue with </div>
                    <div className="grow translate-y-1/2 border-t-[1px] border-slate-400">
                        {' '}
                    </div>
                </div>
                <div className="mb-3 text-center">
                    <button className="text-xl w-full h-9 bg-gray-700 text-white rounded-md">
                        Google Sign In
                    </button>
                    <div
                        id="g_id_onload"
                        data-client_id="860287273531-n6vmem35jtadhea1tlu9hd2ed45l4c6v.apps.googleusercontent.com"
                        data-context="signin"
                        data-ux_mode="popup"
                        data-login_uri="https://api.color4.me"
                        data-auto_prompt="false"
                    ></div>

                    <div
                        className="g_id_signin"
                        data-type="standard"
                        data-shape="rectangular"
                        data-theme="outline"
                        data-text="signin_with"
                        data-size="large"
                        data-locale="en"
                        data-logo_alignment="left"
                    ></div>
                </div>
            </div>
        </div>
    );
}

const Logo = styled(Image)`
    filter: invert(30%);
`;

const Text = styled.p`
    color: rgb(71, 85, 105);
    font-size: 1.25rem;
    font-weight: 600;
    /* font-family: inherit; */
    text-align: center;
`;

function LogoCard() {
    return (
        <Card
            style={{
                margin: 'auto',
                border: 0,
            }}
            cover={
                <Logo
                    preview={false}
                    src={logo}
                    style={{
                        width: 60,
                        margin: 'auto',
                    }}
                />
            }
        >
            <Text>Sign in to your account</Text>
        </Card>
    );
}

export default SignIn;
