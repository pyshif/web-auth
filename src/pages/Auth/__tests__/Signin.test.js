import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import SignIn from '../SignIn';


it('should be users name', () => {
    const { asFragment } = render(<SignIn />);
    // console.log('element :>> ', element);
    expect(asFragment(<SignIn />)).toMatchSnapshot();
});


