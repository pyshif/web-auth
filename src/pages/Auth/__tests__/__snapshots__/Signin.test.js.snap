// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be users name 1`] = `
<DocumentFragment>
  <div
    class="w-full max-w-sm mx-auto"
  >
    <div>
      <div
        class="w-16 mx-auto mb-6"
      >
        <img
          alt=""
          src="[object Object]"
        />
      </div>
      <h1
        class="title font-medium text-3xl text-center mb-3"
      >
        Sign in to your account
      </h1>
      <h2
        class="sub-titile text-md text-center mb-6"
      >
        Or 
        <span
          class="text-slate-600"
        >
          you can 
          <a
            class="underline"
            href="/auth/signup/"
          >
            sign up by here
          </a>
        </span>
      </h2>
    </div>
    <div
      class="shadow-xl bg-white rounded-xl p-8"
    >
      <form>
        <div
          class="mb-3"
        >
          <label
            class="block indent-1 mb-1"
            for="email-address"
          >
            Email address
          </label>
          <input
            class="w-full h-8 px-1 border-[1px] rounded-md"
            id="email-address"
            name="emailAddress"
            type="email"
          />
          <div
            id="email-error-message"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="block indent-1 mb-1"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full h-8 px-1 border-[1px] rounded-md"
            id="password"
            name="password"
            type="password"
          />
          <div
            id="password-error-message"
          />
        </div>
        <div
          class="mb-3 flow-root"
        >
          <a
            class="text-sm text-slate-800 float-right"
            href="/auth/forgot/"
          >
            Forgot your password ?
          </a>
        </div>
        <div
          class="mb-3 text-center"
        >
          <button
            class="text-xl w-full h-9 bg-gray-700 text-white rounded-md"
          >
            Sign in
          </button>
        </div>
      </form>
      <div
        class="relative flex mb-3"
      >
        <div
          class="grow translate-y-1/2 border-t-[1px] border-slate-400"
        >
           
        </div>
        <div
          class="whitespace-normal"
        >
           Or continue with 
        </div>
        <div
          class="grow translate-y-1/2 border-t-[1px] border-slate-400"
        >
           
        </div>
      </div>
      <div
        class="mb-3 text-center"
      >
        <button
          class="text-xl w-full h-9 bg-gray-700 text-white rounded-md"
        >
          Google Sign In
        </button>
        <div
          data-auto_prompt="false"
          data-client_id="860287273531-n6vmem35jtadhea1tlu9hd2ed45l4c6v.apps.googleusercontent.com"
          data-context="signin"
          data-login_uri="https://api.color4.me"
          data-ux_mode="popup"
          id="g_id_onload"
        />
        <div
          class="g_id_signin"
          data-locale="en"
          data-logo_alignment="left"
          data-shape="rectangular"
          data-size="large"
          data-text="signin_with"
          data-theme="outline"
          data-type="standard"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
